python/lsst/pipe/base/butlerQuantumContext.py

```diff
@@ -0,0 +1,204 @@
+                    val = self._get(ref)
+                retVal[name] = val
+            return retVal
+        elif isinstance(dataset, list):
+            return [self._get(x) for x in dataset]
+        elif isinstance(dataset, DatasetRef) or isinstance(dataset, DeferredDatasetRef):
+            return self._get(dataset)
+        else:
+            raise TypeError("Dataset argument is not a type that can be used to get")
+
+    def put(self, values: typing.Union[Struct, typing.List[typing.Any], object],
+            dataset: typing.Union[OutputQuantizedConnection, typing.List[DatasetRef], DatasetRef]):
+        """Puts data into the butler
+
+        Parameters
+        ----------
+        values : `Struct` or `list` of `object` or `object`
+            The data that should be put with the butler. If the type of the dataset
+            is `OutputQuantizedConnection` then this argument should be a `Struct`
+            with corresponding attribute names. Each attribute should then correspond
+            to either a list of object or a single object depending of the type of the
+            corresponding attribute on dataset. I.e. if dataset.calexp is [datasetRef1,
+            datasetRef2] then values.calexp should be [calexp1, calexp2]. Like wise
+            if there is a single ref, then only a single object need be passed. The same
+            restriction applies if dataset is directly a `list` of `DatasetRef` or a
+            single `DatasetRef`.
+        dataset : `OutputQuantizedConnection` or `list` of `lsst.daf.butler.DatasetRef`
+                or `lsst.daf.butler.DatasetRef`
+            This argument may either be an `InputQuantizedConnection` which describes
+            all the inputs of a quantum, a list of `lsst.daf.butler.DatasetRef`, or
+            a single `lsst.daf.butler.DatasetRef`. The function will get and return
+            the corresponding datasets from the butler.
+
+        Raises
+        ------
+        ValueError
+            If a `DatasetRef` is passed to put that is not defined in the quantum object, or
+            the type of values does not match what is expected from the type of dataset.
+        """
+        if isinstance(dataset, OutputQuantizedConnection):
+            if not isinstance(values, Struct):
+                raise ValueError("dataset is a OutputQuantizedConnection, a Struct with corresponding"
+                                 "attributes must be passed as the values to put")
+            for name, refs in dataset:
+                valuesAttribute = getattr(values, name)
+                if isinstance(refs, list):
+                    if len(refs) != len(valuesAttribute):
+                        raise ValueError(f"There must be a object to put for every Dataset ref in {name}")
+                    for i, ref in enumerate(refs):
+                        self._put(valuesAttribute[i], ref)
+                else:
+                    self._put(valuesAttribute, refs)
```

Comment:
I think this could be slightly simplified if `_put()` supported lists.

Reply:
I debated back and forth on this, if it was better to have a simple function that did one thing, or do duplicate the handling the lists and single values. I also debated having an intermediate function to handle lists. Ultimately I decided on a simpler _put, but I don't really have strong feelings.